File: plugins/desktop-excel-mcp/scripts/mcp_stdio_probe.mjs
// 只通过 initialize 与 tools/list 检查 stdio MCP 服务。

import { spawn, spawnSync } from "node:child_process";
import { once } from "node:events";
import { createInterface } from "node:readline";

// 参数：命令行；返回值：解析后的探针设置。
function parse_arguments(argv) {
  const separator_index = argv.indexOf("--");
  if (separator_index < 0 || separator_index === argv.length - 1) {
    throw new Error("missing MCP server command");
  }
  const options = argv.slice(0, separator_index);
  const command = argv.slice(separator_index + 1);
  const read_option = (name, fallback = undefined) => {
    const option_index = options.indexOf(name);
    return option_index >= 0 ? options[option_index + 1] : fallback;
  };
  const name = read_option("--name");
  const cwd = read_option("--cwd");
  const timeout_ms = Number(read_option("--timeout", "45")) * 1000;
  if (!name || !cwd || !Number.isFinite(timeout_ms)) {
    throw new Error("invalid probe arguments");
  }
  return { name, cwd, timeout_ms, command };
}

// 参数：毫秒；返回值：延迟 Promise。
function delay(milliseconds) {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

// 参数：子进程；返回值：无。
async function stop_process_tree(child) {
  if (child.stdin && !child.stdin.destroyed) {
    child.stdin.end();
  }
  if (child.exitCode === null) {
    await Promise.race([once(child, "exit"), delay(2000)]);
  }
  if (child.exitCode === null) {
    if (process.platform === "win32") {
      spawnSync("taskkill.exe", ["/PID", String(child.pid), "/T", "/F"], {
        stdio: "ignore",
        windowsHide: true,
      });
    } else {
      child.kill("SIGKILL");
    }
  }
}

// 参数：探针设置；返回值：进程退出码。
async function probe_server(settings) {
  const [file_path, ...arguments_list] = settings.command;
  const child = spawn(file_path, arguments_list, {
    cwd: settings.cwd,
    windowsHide: true,
    stdio: ["pipe", "pipe", "pipe"],
  });
  child.stdin.setDefaultEncoding("utf8");
  child.stdout.setEncoding("utf8");
  child.stderr.setEncoding("utf8");

  const pending = new Map();
  const stderr_lines = [];
  let fatal_error = null;

  // 参数：错误；返回值：无。
  const fail_pending = (error) => {
    fatal_error = error;
    for (const { reject, timer } of pending.values()) {
      clearTimeout(timer);
      reject(error);
    }
    pending.clear();
  };

  child.on("error", fail_pending);
  child.on("exit", (code) => {
    if (pending.size > 0) {
      fail_pending(new Error(`server exited with code ${code}`));
    }
  });
  child.stderr.on("data", (chunk) => {
    stderr_lines.push(...String(chunk).split(/\r?\n/).filter(Boolean));
    if (stderr_lines.length > 20) {
      stderr_lines.splice(0, stderr_lines.length - 20);
    }
  });

  const reader = createInterface({ input: child.stdout });
  reader.on("line", (line) => {
    if (!line.trim()) {
      return;
    }
    let message;
    try {
      message = JSON.parse(line);
    } catch {
      fail_pending(new Error(`non-JSON stdout: ${line.slice(0, 200)}`));
      return;
    }
    const waiter = pending.get(message.id);
    if (!waiter) {
      return;
    }
    pending.delete(message.id);
    clearTimeout(waiter.timer);
    if (message.error) {
      waiter.reject(new Error(`JSON-RPC error: ${JSON.stringify(message.error)}`));
    } else {
      waiter.resolve(message);
    }
  });

  // 参数：JSON-RPC 消息；返回值：无。
  const send_message = (message) => {
    if (fatal_error) {
      throw fatal_error;
    }
    child.stdin.write(`${JSON.stringify(message)}\n`);
  };

  // 参数：请求消息；返回值：响应 Promise。
  const request = (message) => new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      pending.delete(message.id);
      reject(new Error(`request ${message.id} timed out`));
    }, settings.timeout_ms);
    pending.set(message.id, { resolve, reject, timer });
    try {
      send_message(message);
    } catch (error) {
      clearTimeout(timer);
      pending.delete(message.id);
      reject(error);
    }
  });

  try {
    const initialize_response = await request({
      jsonrpc: "2.0",
      id: 1,
      method: "initialize",
      params: {
        protocolVersion: "2024-11-05",
        capabilities: {},
        clientInfo: { name: "desktop-excel-health-check", version: "0.2.0" },
      },
    });
    send_message({ jsonrpc: "2.0", method: "notifications/initialized", params: {} });
    const tools_response = await request({
      jsonrpc: "2.0",
      id: 2,
      method: "tools/list",
      params: {},
    });
    const tools = tools_response?.result?.tools;
    if (!Array.isArray(tools) || tools.length === 0) {
      throw new Error("tools/list returned no tools");
    }
    const server_name = initialize_response?.result?.serverInfo?.name || settings.name;
    process.stdout.write(`initialize=${server_name};tools=${tools.length}\n`);
    return 0;
  } catch (error) {
    const stderr_tail = stderr_lines.join(" | ").slice(0, 600);
    const suffix = stderr_tail ? `; stderr=${stderr_tail}` : "";
    process.stderr.write(`${settings.name}: ${error.message}${suffix}\n`);
    return 1;
  } finally {
    reader.close();
    await stop_process_tree(child);
  }
}

// 参数：无；返回值：进程退出码。
async function main() {
  const settings = parse_arguments(process.argv.slice(2));
  return probe_server(settings);
}

process.exitCode = await main();

File: plugins/desktop-word-mcp/scripts/mcp_stdio_probe.py
# coding=utf-8
"""只通过 initialize 与 tools/list 检查 stdio MCP 服务。"""

from __future__ import annotations

import argparse
import json
import os
import queue
import subprocess
import sys
import threading
import time
from collections.abc import Sequence


def parse_arguments() -> argparse.Namespace:
    """参数：命令行；返回值：探针设置与服务启动命令。"""
    parser = argparse.ArgumentParser()
    parser.add_argument("--name", required=True)
    parser.add_argument("--timeout", type=float, default=45.0)
    parser.add_argument("--cwd", required=True)
    parser.add_argument("command", nargs=argparse.REMAINDER)
    arguments = parser.parse_args()
    if arguments.command and arguments.command[0] == "--":
        arguments.command = arguments.command[1:]
    if not arguments.command:
        parser.error("missing MCP server command")
    return arguments


def read_stream_lines(stream: object, output_queue: queue.Queue[str | None]) -> None:
    """参数：文本流与输出队列；返回值：无。"""
    try:
        for line in stream:  # type: ignore[union-attr]
            output_queue.put(line.rstrip("\r\n"))
    finally:
        output_queue.put(None)


def collect_stderr(stream: object, lines: list[str]) -> None:
    """参数：stderr 文本流与目标列表；返回值：无。"""
    for line in stream:  # type: ignore[union-attr]
        lines.append(line.rstrip("\r\n"))
        if len(lines) > 20:
            del lines[0]


def send_message(process: subprocess.Popen[str], message: dict) -> None:
    """参数：服务进程与 JSON-RPC 消息；返回值：无。"""
    if process.stdin is None:
        raise RuntimeError("server stdin is unavailable")
    process.stdin.write(json.dumps(message, ensure_ascii=False, separators=(",", ":")) + "\n")
    process.stdin.flush()


def receive_response(
    process: subprocess.Popen[str],
    output_queue: queue.Queue[str | None],
    request_id: int,
    deadline: float,
) -> dict:
    """参数：进程、输出队列、请求 ID 与截止时间；返回值：对应响应。"""
    while time.monotonic() < deadline:
        remaining = max(0.05, deadline - time.monotonic())
        try:
            line = output_queue.get(timeout=min(0.25, remaining))
        except queue.Empty:
            if process.poll() is not None:
                raise RuntimeError(f"server exited with code {process.returncode}")
            continue
        if line is None:
            raise RuntimeError(f"server closed stdout with code {process.poll()}")
        if not line.strip():
            continue
        try:
            message = json.loads(line)
        except json.JSONDecodeError as error:
            raise RuntimeError(f"non-JSON stdout: {line[:200]}") from error
        if message.get("id") != request_id:
            continue
        if "error" in message:
            raise RuntimeError(f"JSON-RPC error: {message['error']}")
        return message
    raise TimeoutError(f"request {request_id} timed out")


def stop_process_tree(process: subprocess.Popen[str]) -> None:
    """参数：探针创建的服务进程；返回值：无。"""
    if process.stdin is not None:
        try:
            process.stdin.close()
        except OSError:
            pass
    try:
        process.wait(timeout=2)
        return
    except subprocess.TimeoutExpired:
        pass
    if os.name == "nt":
        subprocess.run(
            ["taskkill.exe", "/PID", str(process.pid), "/T", "/F"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            check=False,
        )
    else:
        process.kill()
    try:
        process.wait(timeout=5)
    except subprocess.TimeoutExpired:
        process.kill()


def probe_server(name: str, command: Sequence[str], cwd: str, timeout: float) -> int:
    """参数：服务名称、启动命令、目录与超时；返回值：进程退出码。"""
    creation_flags = subprocess.CREATE_NO_WINDOW if os.name == "nt" else 0
    process = subprocess.Popen(
        list(command),
        cwd=cwd,
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        encoding="utf-8",
        errors="replace",
        bufsize=1,
        creationflags=creation_flags,
    )
    output_queue: queue.Queue[str | None] = queue.Queue()
    stderr_lines: list[str] = []
    threading.Thread(target=read_stream_lines, args=(process.stdout, output_queue), daemon=True).start()
    threading.Thread(target=collect_stderr, args=(process.stderr, stderr_lines), daemon=True).start()

    deadline = time.monotonic() + timeout
    try:
        send_message(
            process,
            {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "initialize",
                "params": {
                    "protocolVersion": "2024-11-05",
                    "capabilities": {},
                    "clientInfo": {"name": "desktop-word-health-check", "version": "0.2.0"},
                },
            },
        )
        initialize_response = receive_response(process, output_queue, 1, deadline)
        server_info = initialize_response.get("result", {}).get("serverInfo", {})
        send_message(process, {"jsonrpc": "2.0", "method": "notifications/initialized", "params": {}})
        send_message(process, {"jsonrpc": "2.0", "id": 2, "method": "tools/list", "params": {}})
        tools_response = receive_response(process, output_queue, 2, deadline)
        tools = tools_response.get("result", {}).get("tools", [])
        if not isinstance(tools, list) or not tools:
            raise RuntimeError("tools/list returned no tools")
        print(f"initialize={server_info.get('name') or name};tools={len(tools)}")
        return 0
    except Exception as error:
        stderr_tail = " | ".join(line for line in stderr_lines if line.strip())
        details = f"{name}: {error}"
        if stderr_tail:
            details += f"; stderr={stderr_tail[:600]}"
        print(details, file=sys.stderr)
        return 1
    finally:
        stop_process_tree(process)


def main() -> int:
    """参数：无；返回值：探针退出码。"""
    arguments = parse_arguments()
    return probe_server(arguments.name, arguments.command, arguments.cwd, arguments.timeout)


if __name__ == "__main__":
    raise SystemExit(main())
